client/acl.go

```diff
@@ -48,100 +49,121 @@ type clientACLResolver struct {
 	}
 
-	// Check if this is a management token
-	if token.Type == structs.ACLManagementToken {
-		return acl.ManagementACL, token, nil
-	}
+	var policies []*structs.ACLPolicy
 
-	// Resolve the policy links within the token ACL roles.
-	policyNames, err := c.resolveTokenACLRoles(secretID, token.Roles)
-	if err != nil {
-		return nil, nil, err
-	}
+	// Resolve token policies
+	if token := ident.ACLToken; token != nil {
+		// Check if this is a management token
+		if ident.ACLToken.Type == structs.ACLManagementToken {
+			return acl.ManagementACL, ident, nil
+		}
+
+		// Resolve the policy links within the token ACL roles.
+		policyNames, err := c.resolveTokenACLRoles(secretID, token.Roles)
+		if err != nil {
+			return nil, nil, err
+		}
 
-	// Generate a slice of all policy names included within the token, taken
-	// from both the ACL roles and the direct assignments.
-	policyNames = append(policyNames, token.Policies...)
+		// Generate a slice of all policy names included within the token, taken
+		// from both the ACL roles and the direct assignments.
+		policyNames = append(policyNames, token.Policies...)
 
-	// Resolve the policies
-	policies, err := c.resolvePolicies(token.SecretID, policyNames)
-	if err != nil {
-		return nil, nil, err
+		// Resolve ACL token policies
+		if policies, err = c.resolvePolicies(token.SecretID, policyNames); err != nil {
+			return nil, nil, err
+		}
+	} else {
+		// Resolve policies for workload identities
+		policyArgs := structs.GenericRequest{
+			QueryOptions: structs.QueryOptions{
+				AuthToken: secretID,
+				Region:    c.Region(),
+			},
+		}
+		policyReply := structs.ACLPolicySetResponse{}
+		if err := c.RPC("ACL.GetClaimPolicies", &policyArgs, &policyReply); err != nil {
+			return nil, nil, err
+		}
+		policies = make([]*structs.ACLPolicy, 0, len(policyReply.Policies))
+		for _, p := range policyReply.Policies {
+			policies = append(policies, p)
+		}
 	}
 
 	// Resolve the ACL object
 	aclObj, err := structs.CompileACLObject(c.aclCache, policies)
 	if err != nil {
 		return nil, nil, err
 	}
-	return aclObj, token, nil
+	return aclObj, ident, nil
 }
 
+// TODO(schmichael) rename?
 // resolveTokenValue is used to translate a secret ID into an ACL token with caching
 // We use a local cache up to the TTL limit, and then resolve via a server. If we cannot
 // reach a server, but have a cached value we extend the TTL to gracefully handle outages.
-func (c *Client) resolveTokenValue(secretID string) (*structs.ACLToken, error) {
+func (c *Client) resolveTokenValue(secretID string) (*structs.AuthenticatedIdentity, error) {
 	// Hot-path the anonymous token
 	if secretID == "" {
-		return structs.AnonymousACLToken, nil
+		return &structs.AuthenticatedIdentity{ACLToken: structs.AnonymousACLToken}, nil
 	}
 
 	// Lookup the token entry in the cache
 	entry, ok := c.tokenCache.Get(secretID)
 	if ok {
+		//TODO(schmichael) use a different cache for workload identities since we
```

Comment:
The "expensive" cache of compiled `acl.ACL` objects is shared between tokens and WI because the claims just give us policies that compile down into the same `acl.ACL`. 

Maybe the thing that really wants to happen here is that the `tokenCacheSize` should be larger than the `aclCacheSize`, because it's likely that tokens are in a N:1 ratio with sets of policies. (To use a trivial example, lots of WI's will have a policy set of `[]*structs.ACLPolicy{}` or "no additional policies")

Reply:
Good call, doubled token cache size to 128 to accommodate workload IDs. The TTL is awfully short (30s), and workload ids give us the opportunity to be less naive about caching... but today they're bearer tokens as far as clients are concerned so I'm going to treat them like one.